File: app/src/main/java/com/mycompany/myapp/app/BaseApplicationInitializer.java
package com.mycompany.myapp.app;

import android.app.Application;
import android.content.Intent;

import com.google.android.gms.common.GoogleApiAvailability;
import com.google.android.gms.security.ProviderInstaller;
import com.google.android.gms.security.ProviderInstaller.ProviderInstallListener;
import com.squareup.leakcanary.LeakCanary;

import timber.log.Timber;
import timber.log.Timber.Tree;

public abstract class BaseApplicationInitializer {
    protected final Application application;
    private final Tree logger;

    public BaseApplicationInitializer(
            Application application,
            Tree logger) {
        this.application = application;
        this.logger = logger;
    }

    public void initialize() {
        LeakCanary.install(application);

        Timber.plant(logger);

        upgradeSecurityProvider();
    }

    private void upgradeSecurityProvider() {
        ProviderInstaller.installIfNeededAsync(application, new ProviderInstallListener() {
            @Override
            public void onProviderInstalled() {

            }

            @Override
            public void onProviderInstallFailed(int errorCode, Intent recoveryIntent) {
                GoogleApiAvailability.getInstance().showErrorNotification(application, errorCode);
            }
        });
    }
}

File: app/src/main/java/com/atomicrobot/carbon/app/BaseApplicationInitializer.kt
package com.atomicrobot.carbon.app

import android.app.Application
import android.content.Intent

import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.security.ProviderInstaller
import com.google.android.gms.security.ProviderInstaller.ProviderInstallListener

import timber.log.Timber
import timber.log.Timber.Tree

abstract class BaseApplicationInitializer(
        protected val application: Application,
        private val logger: Tree) {

    open fun initialize() {
        Timber.plant(logger)

        upgradeSecurityProvider()
    }

    private fun upgradeSecurityProvider() {
        ProviderInstaller.installIfNeededAsync(application, object : ProviderInstallListener {
            override fun onProviderInstalled() {

            }

            override fun onProviderInstallFailed(errorCode: Int, recoveryIntent: Intent?) {
                GoogleApiAvailability.getInstance().showErrorNotification(application, errorCode)
            }
        })
    }
}
